File: apps/web/canvas/state/timeline.ts
export type TimelineClip = {
  id: string
  name: string
  subtitle?: string
  volume?: number
  fadeInFrames?: number
  fadeOutFrames?: number
  source: {
    nodeId: string
    kind: 'image' | 'video'
    url: string
    durationFrames: number
    /** 分镜服务端执行装配标识：用于判断本轮是否已装配 */
    runId?: string
    shotId?: string
  }
  inFrame: number
  outFrame: number
}
export type AudioClip = {
  id: string
  name: string
  source: { assetId: string; url: string; durationFrames: number }
  startFrame: number
  inFrame: number
  outFrame: number
  volume: number
  fadeInFrames: number
  fadeOutFrames: number
}
export type Timeline = {
  /** Position of the document's embedded editor in world coordinates. */
  canvas?: { x: number; y: number }
  version: 1
  fps: 30
  width: number
  height: number
  clips: TimelineClip[]
  audioClips?: AudioClip[]
}
export function emptyTimeline(): Timeline {
  return { version: 1, fps: 30, width: 1920, height: 1080, clips: [] }
}
export function timelineDuration(timeline: Timeline): number {
  return timeline.clips.reduce((n, c) => n + c.outFrame - c.inFrame, 0)
}
export function clipAtFrame(timeline: Timeline, frame: number) {
  let start = 0
  for (const clip of timeline.clips) {
    const end = start + clip.outFrame - clip.inFrame
    if (frame >= start && frame < end) {
      return { clip, start, sourceFrame: clip.inFrame + frame - start }
    }
    start = end
  }
  return null
}
export function trimClip(
  timeline: Timeline,
  id: string,
  inFrame: number,
  outFrame: number,
): Timeline {
  const clip = timeline.clips.find((item) => item.id === id)
  if (
    !clip ||
    !Number.isInteger(inFrame) ||
    !Number.isInteger(outFrame) ||
    inFrame < 0 ||
    outFrame <= inFrame ||
    outFrame > clip.source.durationFrames
  ) {
    throw new Error('入点和出点必须在素材时长内，且至少保留一帧')
  }
  return {
    ...timeline,
    clips: timeline.clips.map((item) => (item.id === id ? { ...item, inFrame, outFrame } : item)),
  }
}
export function splitClip(timeline: Timeline, frame: number, newId: string): Timeline {
  const hit = clipAtFrame(timeline, frame)
  if (!Number.isInteger(frame) || !hit || frame === hit.start) {
    throw new Error('请把播放头移到片段内部再分割')
  }
  if (timeline.clips.some((clip) => clip.id === newId)) {
    throw new Error('片段标识重复')
  }
  return {
    ...timeline,
    clips: timeline.clips.flatMap((clip) =>
      clip.id === hit.clip.id
        ? [
            { ...clip, outFrame: hit.sourceFrame },
            { ...clip, id: newId, inFrame: hit.sourceFrame },
          ]
        : [clip],
    ),
  }
}
export function moveClip(timeline: Timeline, id: string, index: number): Timeline {
  const clips = [...timeline.clips]
  const from = clips.findIndex((clip) => clip.id === id)
  if (from < 0 || !Number.isInteger(index) || index < 0 || index >= clips.length) {
    throw new Error('片段位置无效')
  }
  const [clip] = clips.splice(from, 1)
  clips.splice(index, 0, clip)
  return { ...timeline, clips }
}

export function clipVolume(
  clip: Pick<TimelineClip, 'inFrame' | 'outFrame' | 'volume' | 'fadeInFrames' | 'fadeOutFrames'>,
  elapsed: number,
): number {
  const duration = clip.outFrame - clip.inFrame
  const fadeIn = Math.min(clip.fadeInFrames || 0, duration)
  const fadeOut = Math.min(clip.fadeOutFrames || 0, duration)
  return (
    (clip.volume ?? 1) *
    (fadeIn ? Math.max(0, Math.min(1, elapsed / fadeIn)) : 1) *
    (fadeOut ? Math.max(0, Math.min(1, (duration - elapsed) / fadeOut)) : 1)
  )
}

File: apps/api/backend/timeline.py
"""Versioned, frame-based edit decisions. Sources are snapshots, independent of canvas nodes."""

from copy import deepcopy
from math import isfinite


def validate_timeline(raw):
    if not isinstance(raw, dict) or raw.get("version") != 1 or raw.get("fps") != 30:
        raise ValueError("时间线版本或帧率不受支持")
    if "canvas" in raw:
        position = raw["canvas"]
        if not isinstance(position, dict) or any(
            type(position.get(key)) not in (int, float)
            or not isfinite(position[key])
            or abs(position[key]) > 1_000_000
            for key in ("x", "y")
        ):
            raise ValueError("时间线画布位置无效")
    for key in ("width", "height"):
        if type(raw.get(key)) is not int or not 64 <= raw[key] <= 4096:
            raise ValueError("时间线画幅无效")
    clips = raw.get("clips")
    if not isinstance(clips, list) or len(clips) > 2000:
        raise ValueError("时间线片段格式错误或超过 2000 项")
    ids = set()
    for clip in clips:
        if not isinstance(clip, dict):
            raise ValueError("时间线片段格式错误")
        ident = clip.get("id")
        if not isinstance(ident, str) or not ident or len(ident) > 100 or ident in ids:
            raise ValueError("时间线片段标识无效或重复")
        ids.add(ident)
        if not isinstance(clip.get("name"), str) or len(clip["name"]) > 500:
            raise ValueError("时间线片段名称无效")
        if "subtitle" in clip and (
            not isinstance(clip["subtitle"], str) or len(clip["subtitle"]) > 4000
        ):
            raise ValueError("片段字幕无效或超过 4000 字")
        volume = clip.get("volume", 1)
        if type(volume) not in (int, float) or not 0 <= volume <= 1:
            raise ValueError("片段音量必须在 0 到 1 之间")
        for fade in ("fadeInFrames", "fadeOutFrames"):
            if type(clip.get(fade, 0)) is not int or not 0 <= clip.get(fade, 0) <= 108000:
                raise ValueError("音频淡入淡出时长无效")
        src = clip.get("source")
        if not isinstance(src, dict) or src.get("kind") not in ("image", "video"):
            raise ValueError("时间线素材类型无效")
        if not isinstance(src.get("nodeId"), str):
            raise ValueError("时间线来源节点无效")
        url = src.get("url")
        if (
            not isinstance(url, str)
            or not url
            or not (
                (url.startswith("/") and not url.startswith("//"))
                or url.startswith(("https://", "http://", "data:image/"))
            )
        ):
            raise ValueError("时间线素材地址无效")
        values = [clip.get("inFrame"), clip.get("outFrame"), src.get("durationFrames")]
        if (
            any(type(v) is not int for v in values)
            or not 0 <= values[0] < values[1] <= values[2] <= 108000
        ):
            raise ValueError("时间线入出点超出素材时长")
    validate_audio_clips(raw.get("audioClips", []))
    return deepcopy(raw)


def timeline_srt(timeline):
    def stamp(frame):
        milliseconds = (frame * 1000 + 15) // 30
        seconds, ms = divmod(milliseconds, 1000)
        minutes, sec = divmod(seconds, 60)
        hours, minute = divmod(minutes, 60)
        return f"{hours:02}:{minute:02}:{sec:02},{ms:03}"

    import re

    frame = 0
    cues = []
    for clip in timeline["clips"]:
        end = frame + clip["outFrame"] - clip["inFrame"]
        text = re.sub(
            r"\n\s*\n", "\n", clip.get("subtitle", "").replace("\r\n", "\n").replace("\r", "\n")
        ).strip()
        if text:
            cues.append(f"{len(cues) + 1}\n{stamp(frame)} --> {stamp(end)}\n{text}")
        frame = end
    return "\n\n".join(cues) + "\n" if cues else ""


def validate_audio_clips(clips):
    import re

    if not isinstance(clips, list) or len(clips) > 64:
        raise ValueError("音频片段不能超过 64 项")
    ids = set()
    for clip in clips:
        if not isinstance(clip, dict):
            raise ValueError("音频片段格式无效")
        ident = clip.get("id")
        if not isinstance(ident, str) or not ident or len(ident) > 100 or ident in ids:
            raise ValueError("音频片段标识无效或重复")
        ids.add(ident)
        if not isinstance(clip.get("name"), str) or len(clip["name"]) > 500:
            raise ValueError("音频名称无效")
        source = clip.get("source")
        if not isinstance(source, dict) or not isinstance(source.get("assetId"), str):
            raise ValueError("音频资产无效")
        asset = source["assetId"]
        if (
            not re.fullmatch(r"[0-9a-f]{8}(?:-[0-9a-f]{4}){3}-[0-9a-f]{12}", asset)
            or source.get("url") != f"/assets/{asset}/original.wav"
        ):
            raise ValueError("音频来源必须是资产库 WAV")
        start, inside, outside, duration = (
            clip.get("startFrame"),
            clip.get("inFrame"),
            clip.get("outFrame"),
            source.get("durationFrames"),
        )
        if (
            any(type(v) is not int for v in (start, inside, outside, duration))
            or not 0 <= inside < outside <= duration <= 108000
            or start < 0
            or start + outside - inside > 108000
        ):
            raise ValueError("音频起始位置或裁剪范围无效")
        volume = clip.get("volume")
        if type(volume) not in (int, float) or not 0 <= volume <= 1:
            raise ValueError("音频音量无效")
        for field in ("fadeInFrames", "fadeOutFrames"):
            if type(clip.get(field)) is not int or not 0 <= clip[field] <= 108000:
                raise ValueError("音频淡化时长无效")
